### ISSUE TITLE
Reconsider propagating only the inner exception on Control.Invoke

### ISSUE THREAD

**Issue by tbolon** [dotnet/winforms#5352]
**Is your feature request related to a problem? Please describe.**

See this [Stackoverflow comment which describes the current behavior](https://stackoverflow.com/a/28055196/8731)

Currently, the Control.Invoke method only propagates the innermost exception to the caller.
There are multiple justifications from microsoft in the comments (from 2015), and one of the is the "breaking change"-one.
I would like to raise again this problem, and create a discussion around it.

For me, this behavior is misleading and not documented, and can cause headaches about the loss of exception details when InvokeRequired/Invoke code is used. Depending on how the code is executed, the Invoke code path could cause the detail of the exception to be loss.

In my case, I have frequent code which ensure that the current thread is the UI one, and if not, uses Control.Invoke to call the method in the correct context.

Furthermore, if your code is based on [WindowsFormsSynchronizationContext](https://github.com/dotnet/winforms/blob/81774abbcbb885fe906f1c0f813e8175eef8705c/src/System.Windows.Forms/src/System/Windows/Forms/WindowsFormsSynchronizationContext.cs#L87), there is no way to use a workaround and preserve the inner exceptions.

**Describe the solution you'd like and alternatives you've considered**

The primary reason of the removal of exception details was to ensure that the winform default exception dialog could display the most useful exception to the user.

A first solution could be to change the code of [ThreadContext.OnThreadException](https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/src/System/Windows/Forms/Application.ThreadContext.cs#L871) to call ThreadExceptionDialog with the the innermost exception details, then let the Control.Invoke method raise the exception with all details.  
This is a breaking change because [Application.ThreadException](https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/src/System/Windows/Forms/Application.cs#L771) will now raise

A second solution could to to change the code of [ThreadContext.OnThreadException](https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/src/System/Windows/Forms/Application.ThreadContext.cs#L871) but now retrieve the innermost exception details for both the ThreadException event and ThreadExceptionDialog, and let Control.Invoke preserve all exception details. This is also a breaking change, but now only in the scope of the Control.Invoke behavior.

And, at least, if no solution is accepted, the current behavior of Control.Invoke/BeginInvoke should be better described in documentation. No breaking change in this case :)

**Comment by tbolon** [comment]
Sorry, I initially closed this issue because I didn't test if this behavior was present in latest .net core.

I can now confirm that the behavior is still the same in .NET 6.

**Comment by tbolon** [comment]
I am still trying to understand how exception are propagated when using Control.Invoke/BeginInvoke.

Please ignore this issue for now, I will update it with a solution when I find one.

**Comment by tbolon** [comment]
Based on my investigation, there is no trivial way to throw the detailed exception AND preserve the current behavior of showing only the innermost exception in Application.ThreadException and the ThreadExceptionDialog.

I would like to keep my suggestion opened for discussion as I see no real interest in having Control.Invoke only keeping the innermost exception, or at least suggest some documentation changes.

**Comment by RussKie** [comment]
Thank you for raising the issue. It is also late in .NET 6.0 dev cycle, we need to have especially strong reasons to make breaking changes now.
Please bear with us, this is a complex topic which requires time to sit down, investigate and think it through. The team is quite pre-occupied making sure .NET 6.0 and the designer are as good as they can be. We likely be revisiting this discussion closer/after .NET 6.0 goes GA.

**Comment by tbolon** [comment]
No problem, I did not intend that this change should be considered for .NET 6.0.

I just encountered this problem in my app, and found the problem interesting.
And since a lot have changed since 2015, I found useful to raise this question again.

For sure, take your time polishing the release of .NET 6.0, and take some leave if you can :)

**Comment by JeremyKuhne** [comment]
@KlausLoeffelmann this is somewhat related to the async stuff you're looking at. Might be a useful option or overload perhaps?

**Comment by KlausLoeffelmann** [comment]
Might be. This should be included in some parent epic issue to keep track of all the async-context issues, IMHO. @merriemcgaw FYI.

**Comment by merriemcgaw** [comment]
@KlausLoeffelmann Yes! I think that's a great idea. As we get deeper into .NET 9 planning this is one place where I know we need to invest.

**Comment by KlausLoeffelmann** [comment]
In that, we should also include the community API suggestion of automatically marshalling a ViewModel Property Change on a non-UI thread to the UI-thread.

Are you setting that Epic up, or do you want me to do that?

**Comment by JeremyKuhne** [comment]
@KlausLoeffelmann can you update as appropriate based on other async work?

**Comment by tbolon** [comment]
Hello all. I just wanted to say that this problem has hit us back as we are transitioning to using async UI Events.

We use Application.ThreadException to diagnose and report issues in our app, but as of now, only the leaf exception is reported as soon as we use async void event hander:

```csharp
        // this will report the root exception in Application.ThreadException
        private void button2_Click(object sender, EventArgs e)
        {
            try
            {
                throw new InvalidOperationException("Leaf Exception");
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException("Root Exception", ex);
            }
        }

        // this will only report the leaf exception...
        private async void button1_Click(object sender, EventArgs e)
        {
            try
            {
                throw new InvalidOperationException("Leaf Exception");
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException("Root Exception", ex);
            }
        }
```

We lose a ton of contextuel information when our code run is an async path, and are not able to recover from transient errors because of this. I found no other workaround than try/catching all our async calls to report the full exception.

For now, we stopped migrating to async events because of this, sadly.
